### ISSUE TITLE
FSE - Template meta information and post type support do not work

### ISSUE THREAD

**Issue by rufus87** [WordPress/gutenberg#31704]
## Description
Custom template configuration in `experimental-theme.json` file does not work neither in default themes nor in custom ones.

## Step-by-step reproduction instructions
1. Create `my-custom-template.html` file within current active theme's "block-templates" folder
2. Add configuration from `Code snippet` part to active theme's `experimental-theme.json` file
3. Go to admin panel and go to "Posts -> Add New"
4. In the opened screen open "Template dropdown"

## Expected behaviour
1. `Template` dropdown should not contain above mentioned `my-custom-template` as an option because it should not be available for any post type but `page`.
2. The template name must be read as `The template title` instead of `my-custom-template` based on `experimental-theme.json` file configuration

## Actual behaviour
1. Template is available in all post types: `postTypes` restriction does not work
2. Template name is read as `my-custom-template` instead of `The template title`

## Screenshots or screen recording (optional)
![image](https://user-images.githubusercontent.com/10531492/117814435-5ad8af80-b275-11eb-93d0-e586514dbc13.png)
![image](https://user-images.githubusercontent.com/10531492/117814484-69bf6200-b275-11eb-8490-7d695299cce1.png)
![image](https://user-images.githubusercontent.com/10531492/117814509-7479f700-b275-11eb-929c-1b962ad6408b.png)

## Code snippet (optional)
```
...
"customTemplates": [
    {
      "name": "my-custom-template",
      "title": "The template title",
      "postTypes": [
        "page"
      ]
    }
  ]
...
```

## WordPress information
- WordPress version: 5.7.1
- Gutenberg version: 10.5.4
- Are all plugins except Gutenberg deactivated? Yes
- Are you using a default theme (e.g. Twenty Twenty-One)? No

## Device information
- Device: Desktop
- Operating system: Windows 10
- Browser: Chrome 90.0.4430.93 (Official Build) (64-bit)

**Comment by rufus87** [comment]
Hi there, 

Many thanks for your feedback, however it is not visible here.

Seams like you did not understand the issue correctly. The template appears in "Templates" dropdown in any post type, ie: "postTypes" restriction in `experimental-theme.json` file does not work. Right now "postTypes" is set to `['page']`, but template is available for selecting in `Templates` dropdown when I try to add `post` - not `page`, as well as the template `filename` appears in dropdown instead of the configured `title`. There is no difference how template file is named, does it have `page-` prefix or not.

Thanks

**Comment by Aljullu** [comment]
I can reproduce this as well. For example, if I create a `single-product.html` template and then I add:

```JSON
"customTemplates": [
	{
		"name": "single-product",
		"title": "Product page",
		"postTypes": [
			"product"
		]
	}
]
```
to my `theme.json`, the `single-product` template is still available in pages and posts. Also, the `title` property is ignored and the template is listed as `single-product` instead of `Product page`.

It seems to be related to this `TODO` comment in the code:

https://github.com/WordPress/gutenberg/blob/9d382e4d7f8b1afba2e29fb66e98b95d5ba09c7a/lib/full-site-editing/page-templates.php#L21

I think this was regressed in #30438, specifically in these changes: https://github.com/WordPress/gutenberg/commit/7c134eb66f1e5ed67db646e19234d9edf6e3090e#diff-603069a6df3015f09478e5ed97c90a404f2e17fb29dc7af5467dac8131acc1c6. cc @youknowriad for awareness and in case you think this is something that needs to be prioritized.

**Comment by Mamaduka** [comment]
Created PR that should fix the `customTemplates` title issue - #35592.

I'm working on excluding templates based on "postTypes" PR should be ready shortly.
